src/main/java/com/google/cloud/dataproc/templates/main/DataProcTemplate.java

```diff
@@ -15,86 +15,116 @@
+    PropertyUtil.registerProperties(properties);
+
+    TemplateName templateName;
     try {
-      template = TemplateName.valueOf(templateName);
+      templateName = TemplateName.valueOf(templateNameString.trim().toUpperCase());
     } catch (IllegalArgumentException ex) {
-      LOGGER.error("Unexpected template name :{}", templateName, ex);
-      throw ex;
+      throw new IllegalArgumentException(String.format("Unexpected template name: %s", templateNameString), ex);
     }
-    LOGGER.info("Running job for {}", template);
-    runSparkJob(template);
+    LOGGER.info("Running job for {}", templateName);
+    runSparkJob(templateName, cmd.getArgs());
   }
 
+  static final Map<TemplateName, Function<String[], BaseTemplate>> templateFactories =
+      ImmutableMap.<TemplateName, Function<String[], BaseTemplate>>builder()
+          .put(TemplateName.WORDCOUNT, (args) -> new WordCount())
+          .put(TemplateName.HIVETOGCS, (args) -> new HiveToGCS())
+          .put(TemplateName.HIVETOBIGQUERY, (args) -> new HiveToBigQuery())
+          .put(TemplateName.PUBSUBTOBQ, (args) -> new PubSubToBQ())
+          .put(TemplateName.GCSTOBIGQUERY, (args) -> new GCStoBigquery())
+          .put(TemplateName.S3TOBIGQUERY, (args) -> new S3ToBigQuery())
+          .put(TemplateName.SPANNERTOGCS, (args) -> new SpannerToGCS())
+          .put(TemplateName.GENERAL, GeneralTemplate::of)
+          .build();
+
   /**
    * Run spark job for template.
    *
-   * @param template name of the template.
+   * @param templateName name of the template to execute.
    */
-  static void runSparkJob(TemplateName template) {
+  static void runSparkJob(TemplateName templateName, String[] args) {
     LOGGER.debug("Start API runSparkJob");
-
-    switch (template) {
-      case WORDCOUNT:
-        {
-          new WordCount().runTemplate();
-        }
-        break;
-      case HIVETOGCS:
-        {
-          new HiveToGCS().runTemplate();
-        }
-        break;
-      case PUBSUBTOBQ:
-        {
-          new PubSubToBQ().runTemplate();
-        }
-        break;
-      case HIVETOBIGQUERY:
-        {
-          new HiveToBigQuery().runTemplate();
-        }
-        break;
-      case GCSTOBIGQUERY:
-        {
-          new GCStoBigquery().runTemplate();
-        }
-        break;
-      case SPANNERTOGCS:
-        {
-          new SpannerToGCS().runTemplate();
-        }
-        break;
-      case S3TOBIGQUERY:
-        {
-          new S3ToBigQuery().runTemplate();
-        }
-        break;
-      default:
-        throw new IllegalArgumentException("Unexpected template name :" + template);
+    BaseTemplate template;
+    if (templateFactories.containsKey(templateName)) {
+      template = templateFactories.get(templateName).apply(args);
+    } else {
+      throw new IllegalArgumentException(
+          String.format("Unexpected template name: %s", templateName));
     }
-
+    template.runTemplate();
```

Comment:
The templates still have to start their own spark session because that's too big a change for this PR, but this would be the perfect place to start, configure and then pass in the spark session to the runTemplate method, to improve consistency and readability of the templates.